### PR TITLE
feat: add support for DS docs system tags

### PR DESCRIPTION
**Pull request by fontanierh**
- add the concept of "system tags"
- "system tags" are transparent to users, excluded from core API responses
- "upsert" by default does not modify system tags
- "system tags" are prefixed by `__DUST`

### CODE REVIEW

File: core/src/data_sources/data_source.rs [hunk 1]
```diff
@@ -1192,6 +1246,7 @@ pub async fn cmd_upsert(
             tags,
             source_url,
             text,
+            true, // preserve system tags
```
Comment (fontanierh) [inline core/src/data_sources/data_source.rs:1249]: @spolu wondering what the best behaviour is for CMD. I would be tempted to return / overwrite all tags in CMD as it's really a different use-case. WDYT ?

File: core/src/data_sources/data_source.rs [hunk 2]
```diff
@@ -1264,7 +1319,10 @@ pub async fn cmd_retrieve(data_source_id: &str, document_id: &str) -> Result<()>
         None => Err(anyhow!("Data source `{}` not found", data_source_id))?,
     };
 
-    let d = match ds.retrieve(Box::new(store.clone()), document_id).await? {
+    let d = match ds
+        .retrieve(Box::new(store.clone()), document_id, true)
```
Comment (fontanierh) [inline core/src/data_sources/data_source.rs:1323]: Same here, tempted to use `false`

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:1249]: no worries about `cmd` we will kill it with fire soon whatever works

File: core/src/data_sources/data_source.rs [hunk 3]
```diff
@@ -1320,7 +1378,7 @@ pub async fn cmd_list(data_source_id: &str) -> Result<()> {
     let project = Project::new_from_id(1);
 
     let r = store
-        .list_data_source_documents(&project, data_source_id, None)
+        .list_data_source_documents(&project, data_source_id, None, true)
```
Comment (fontanierh) [inline core/src/data_sources/data_source.rs:1381]: and here as well

File: core/src/data_sources/data_source.rs (hunk 2 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:1323]: no biggy

File: core/src/data_sources/data_source.rs (hunk 3 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:1381]: no biggy

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (fontanierh) [reply core/src/data_sources/data_source.rs:1249]: Ok 👌

File: core/src/stores/store.rs [hunk 4]
```diff
@@ -111,6 +111,7 @@ pub trait Store {
         project: &Project,
         data_source_id: &str,
         limit_offset: Option<(usize, usize)>,
+        remove_system_tags: bool,
```
Comment (spolu) [inline core/src/stores/store.rs:114]: Do we really need to go down to the `store` level? Seems easier to handle this entirely from `data_source.rs` ?

**Review by spolu — COMMENTED** [review]
Left one comment as it seems easier to contain that abstraction to `data_source.rs` intead of leaking it at the store level?

File: core/src/data_sources/data_source.rs [hunk 5]
```diff
@@ -392,9 +393,51 @@ impl DataSource {
...
+        {
+            Err(anyhow!(
+                "preserve_system_tags=true is not allowed if `tags` contains a string starting with \"{}\"",
+                DATA_SOURCE_DOCUMENT_SYSTEM_TAG_PREFIX
+            ))?;
+        }
+
         let store = store.clone();
 
+        let additional_tags = if preserve_system_tags {
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:412]: For clarity `current_system_tags` ?

File: core/src/data_sources/data_source.rs [hunk 6]
```diff
@@ -392,9 +393,51 @@ impl DataSource {
         tags: &Vec<String>,
         source_url: &Option<String>,
         text: &str,
+        preserve_system_tags: bool,
     ) -> Result<Document> {
+        // disallow preserve_system_tags=true if tags contains a string starting with the system tag prefix
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:398]: Maybe add a sentence to the comment: this guarantees that we don't have duplicate system tags when preserving system tags?

Comment (fontanierh) [reply core/src/data_sources/data_source.rs:398]: Adding ! Also, it prevents people from upserting with `__DUST*` tags from the API / UI

File: core/src/stores/store.rs (hunk 4 above)
Comment (spolu) [reply core/src/stores/store.rs:114]: Ah gotcha :+1: OK LGTM :+1:

**Comment by fontanierh** [comment]
Comments addressed @spolu

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**